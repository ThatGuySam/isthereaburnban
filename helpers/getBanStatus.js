const is = require('./is')
const states = require('./states')
const getLocationInfo = require('./getLocationInfo')
const getMessage = require('./messages')

module.exports = async (geolocation) => {
  
  let locationName
  let stateName
  let googleCountyName
  let countyName
  
  if (is.propertyDefined(geolocation, 'address')) {
    locationName = `${geolocation.address.city}, ${geolocation.address.stateCode}`
    stateName = geolocation.address.state
    stateCode = geolocation.address.stateCode
    googleCountyName = geolocation.address.region
    countyName = googleCountyName.toLowerCase().replace("county", "").trim()
  } else {
    const location = await getLocationInfo(geolocation)
    
    locationName = `${location.city}, ${location.administrativeLevels.level1short}`
    stateName = location.administrativeLevels.level1long
    stateCode = location.administrativeLevels.level1short
    googleCountyName = location.administrativeLevels.level2long
    countyName = googleCountyName.toLowerCase().replace("county", "").trim()
  }

  const state = states(stateCode)
  
  // Check if state is supported
  if (state === null) return {
    key: 'stateNotSupported',
    ...getMessage('stateNotSupported', {name: locationName})
  }
  
  const bans = await state.getBans()
  
  const county = bans.filter(function( county ) {
    return county.name.toLowerCase() == countyName
  })[0]
  
  const key = (county.inEffect) ? 'yes' : 'no'
  const locationResponse = {
    name: locationName,
    county: {
      ...county,
      longName: googleCountyName
    }
  }
  
  const status = getMessage(key, locationResponse)
  
  return {
    key: key,
    location: locationResponse,
    ...status
  }
}