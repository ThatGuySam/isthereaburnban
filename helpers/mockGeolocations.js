const is = require('is_js')

module.exports = (key) => {
  
  const geolocations = {
    'brokenarrow': {
      coords: {
        accuracy: 5,
        altitude: 0,
        altitudeAccuracy: -1,
        heading: -1,
        latitude: 36.7299286,
        longitude: -102.517822,
        speed: -1
      },
      timestamp: 1528944033344.24
    },
    'boisecity': {
      coords: {
        accuracy: 5,
        altitude: 0,
        altitudeAccuracy: -1,
        heading: -1,
        latitude: 36.7297121,
        longitude: -102.5130764,
        speed: -1
      },
      timestamp: 1528944033344.24
    },
    // Texas
    'elpaso': {
      coords: {
        accuracy: 5,
        altitude: 0,
        altitudeAccuracy: -1,
        heading: -1,
        latitude: 31.7688106,
        longitude: -106.4436692,
        speed: -1
      },
      timestamp: 1528944033344.24
    },
    'canadian': {
      coords: {
        accuracy: 5,
        altitude: 0,
        altitudeAccuracy: -1,
        heading: -1,
        latitude: 35.7297121,
        longitude: -100.5130764,
        speed: -1
      },
      timestamp: 1528944033344.24
    }
  }
  
  const hasLocation = is.propertyDefined(geolocations, key)
  
  // Return null if there's no state
  if (!hasLocation) {
    console.log('That location is not defined', key)
    return null
  }
  
  return geolocations[key]
}