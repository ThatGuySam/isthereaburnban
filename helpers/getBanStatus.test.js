import getBanStatus from './getBanStatus'
import mockGeolocations from './mockGeolocations'

describe('Check Bans', () => {
  
  const burnBanStructure = {
    'backgroundColor': expect.any(String),
    'key': expect.any(String),
    'message': expect.any(String),
    'word': expect.any(String)
  }

  it('Can Check El Paso', async () => {
    const geolocation = mockGeolocations('elpaso')
    const burnBans = await getBanStatus(geolocation)
  
    expect(burnBans).toMatchObject(burnBanStructure)
  })
  
  it('Can Check Boise City', async () => {
    const geolocation = mockGeolocations('boisecity')
    const burnBans = await getBanStatus(geolocation)
  
    expect(burnBans).toMatchObject(burnBanStructure)
  })
  
  it('Can Check Broken Arrow', async () => {
    const geolocation = mockGeolocations('brokenarrow')
    const burnBans = await getBanStatus(geolocation)
  
    expect(burnBans).toMatchObject(burnBanStructure)
  })
  
  it('Can Handle Missing County', async () => {
    const geolocation = mockGeolocations('canadian')
    const burnBans = await getBanStatus(geolocation)
    const burnBanWithNoCounty = {
      ...burnBanStructure,
      'key': 'noCounty'
    }
  
    expect(burnBans).toMatchObject(burnBanWithNoCounty)
  })

})