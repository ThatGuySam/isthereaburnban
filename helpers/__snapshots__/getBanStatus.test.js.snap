// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Bans Can Check Broken Arrow: [object Object] 1`] = `
Object {
  "backgroundColor": "#f34f49",
  "key": "yes",
  "location": Object {
    "county": Object {
      "id": 26,
      "inEffect": true,
      "longName": "Cimarron County",
      "name": "Cimarron",
      "status": "Governor's Burn Ban",
    },
    "name": "Boise City, OK",
  },
  "message": "There is a ban in effect for Boise City, OK",
  "word": "Yes",
}
`;

exports[`Check Bans Can Check El Paso: [object Object] 1`] = `
Object {
  "backgroundColor": "#f34f49",
  "key": "yes",
  "location": Object {
    "county": Object {
      "id": 70,
      "inEffect": true,
      "longName": "El Paso County",
      "name": "El Paso",
      "status": "Active",
    },
    "name": "El Paso, TX",
  },
  "message": "There is a ban in effect for El Paso, TX",
  "word": "Yes",
}
`;

exports[`Check Bans Can Handle Missing County: [object Object] 1`] = `
Object {
  "backgroundColor": "#123646",
  "button": Object {
    "label": "Texas Forest Service",
    "url": "http://texasforestservice.tamu.edu/TexasBurnBans/",
  },
  "key": "noCounty",
  "message": "There was problem getting your county. Try checking your state's burn ban site. ",
  "word": "D'oh",
}
`;
